File: app/src/main/java/com/style/base/BaseRecyclerViewAdapter.java
package com.style.base;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.style.manager.LogManager;
import com.style.manager.ToastManager;
import com.style.utils.CommonUtil;
import com.style.utils.DeviceInfoUtil;

import java.sql.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by XiaJun on 2015/7/2.
 */
public abstract class BaseRecyclerViewAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    protected String TAG = getClass().getSimpleName();

    public Context mContext;
    public LayoutInflater mInflater;
    public ArrayList<T> list;
    private OnItemClickListener onItemClickListener;
    private OnItemLongClickListener onItemLongClickListener;

    public BaseRecyclerViewAdapter(Context context, ArrayList<T> dataList) {
        this.list = dataList;
        this.mContext = context;
        mInflater = LayoutInflater.from(context);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }
    public Context getContext() {
        return mContext;
    }
    public ArrayList<T> getList() {
        return list;
    }

    public void setData(ArrayList<T> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    public void addData(T e) {
        this.list.add(e);
        notifyDataSetChanged();
    }

    public void addData(ArrayList<T> list) {
        this.list.addAll(list);
        notifyDataSetChanged();
    }

    public T getData(int position) {
        return list.get(position);
    }

    public void removeData(int position) {
        this.list.remove(position);
        notifyItemRemoved(position);
    }

    public void removeData(T e) {
        this.list.remove(e);
        notifyDataSetChanged();
    }

    public void updateData(int position, T e) {
        list.set(position, e);
        notifyItemChanged(position);
    }

    public void clearData() {
        this.list.clear();
        notifyDataSetChanged();
    }

    public void setOnItemClickListener(RecyclerView.ViewHolder holder, final int position) {
        if (onItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (onItemClickListener != null)
                        onItemClickListener.onItemClick(position, getData(position));
                }
            });
        }
    }

    public void setOnItemLongClickListener(RecyclerView.ViewHolder holder, final int position) {
        if (onItemLongClickListener != null) {
            holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    if (onItemLongClickListener != null)
                        onItemLongClickListener.onItemLongClick(v, position, getData(position));
                    return true;
                }
            });
        }
    }

    public void setOnItemClickListener(OnItemClickListener mListener) {
        if (mListener != null)
            this.onItemClickListener = mListener;
    }

    public interface OnItemClickListener<T> {
        void onItemClick(int position, T data);
    }

    public void setOnItemLongClickListener(OnItemLongClickListener mListener) {
        if (mListener != null)
            this.onItemLongClickListener = mListener;
    }

    public interface OnItemLongClickListener<T> {
        void onItemLongClick(View itemView, int position, T data);
    }

    public void showToast(String str) {
        ToastManager.showToast(mContext, str);
    }

    public void showToast(int resId) {
        ToastManager.showToast(mContext, resId);
    }

    public void logE(String tag, String msg) {
        LogManager.logE(tag, msg);
    }

    protected void setText(TextView textView, String str) {
        CommonUtil.setText(textView, str);
    }

    protected void setText(TextView textView, int strId) {
        setText(textView, mContext.getString(strId));
    }

    protected int dp2px(float dpValue) {
        return DeviceInfoUtil.dp2px(mContext, dpValue);
    }

}

File: app/src/main/java/com/style/base/BaseRecyclerViewAdapter.kt
package com.style.base

import android.content.Context
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.style.toast.ToastManager

import com.style.utils.LogManager
import com.style.utils.DeviceInfoUtil

import java.util.ArrayList

/**
 * Created by XiaJun on 2015/7/2.
 */
abstract class BaseRecyclerViewAdapter<T> : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    protected var TAG = javaClass.simpleName

    private var context: Context
    var list: ArrayList<T>

    private var mInflater: LayoutInflater
    private var onItemClickListener: OnItemClickListener<T>? = null
    private var onItemLongClickListener: OnItemLongClickListener<T>? = null

    constructor(context: Context?, list: ArrayList<T>) : super() {
        this.context = context!!
        this.list = list
        this.mInflater = LayoutInflater.from(context)
    }

    fun getContext(): Context {
        return context
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun <T : ViewDataBinding> getBinding(@LayoutRes layoutId: Int, parent: ViewGroup?): T {
        return DataBindingUtil.bind(mInflater.inflate(layoutId, parent, false))!!
    }

    fun setData(list: ArrayList<T>) {
        this.list = list
        notifyDataSetChanged()
    }

    fun addData(e: T) {
        this.list.add(e)
        notifyDataSetChanged()
    }

    fun addData(list: ArrayList<T>) {
        this.list.addAll(list)
        notifyDataSetChanged()
    }

    fun getData(position: Int): T {
        return list[position]
    }

    fun removeData(position: Int) {
        this.list.removeAt(position)
        notifyItemRemoved(position)
    }

    fun removeData(e: T) {
        this.list.remove(e)
        notifyDataSetChanged()
    }

    fun updateData(position: Int, e: T) {
        list[position] = e
        notifyItemChanged(position)
    }

    fun clearData() {
        this.list.clear()
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(holder: View, position: Int) {
        if (onItemClickListener != null) {
            holder.setOnClickListener {
                onItemClickListener?.onItemClick(position, getData(position))
            }
        }
    }

    fun setOnItemLongClickListener(holder: View, position: Int) {
        if (onItemLongClickListener != null) {
            holder.setOnLongClickListener { v ->
                onItemLongClickListener?.onItemLongClick(v, position, getData(position))
                true
            }
        }
    }

    fun setOnItemClickListener(mListener: OnItemClickListener<T>?) {
        if (mListener != null)
            this.onItemClickListener = mListener
    }

    fun getOnItemClickListener(): OnItemClickListener<T>? {
        return this.onItemClickListener
    }

    interface OnItemClickListener<T> {
        fun onItemClick(position: Int, data: T)
    }

    fun setOnItemLongClickListener(mListener: OnItemLongClickListener<T>?) {
        if (mListener != null)
            this.onItemLongClickListener = mListener
    }

    interface OnItemLongClickListener<T> {
        fun onItemLongClick(itemView: View, position: Int, data: T)
    }

    fun showToast(str: CharSequence) {
        ToastManager.showToast(context, str)
    }

    fun showToast(@StringRes resId: Int) {
        showToast(context.getText(resId))
    }

    fun logE(tag: String, msg: String) {
        LogManager.logE(tag, msg)
    }

    protected fun dp2px(dpValue: Float): Int {
        return DeviceInfoUtil.dp2px(context, dpValue)
    }

}